File: apps/rider/lib/core/iraqi_phone.dart
/// Iraqi (+964) mobile phone handling — mirrors the backend
/// `normalizeIraqiPhone` (services/api/src/common/phone.util.ts) so client-side
/// validation matches the server.
///
/// Iraqi mobiles are 10 national digits starting with `7`; canonical E.164 is
/// `+9647XXXXXXXXX`.
abstract final class IraqiPhone {
  static final RegExp _national = RegExp(r'^7\d{9}$');
  static final RegExp _strip = RegExp(r'[\s\-().]');

  /// Returns the canonical `+9647XXXXXXXXX` form, or `null` if [input] is not a
  /// valid Iraqi mobile number.
  static String? normalize(String input) {
    var s = input.trim().replaceAll(_strip, '');

    // 00 international prefix → +
    if (s.startsWith('00')) s = '+${s.substring(2)}';

    String national;
    if (s.startsWith('+964')) {
      national = s.substring(4);
    } else if (s.startsWith('964')) {
      national = s.substring(3);
    } else if (s.startsWith('0')) {
      national = s.substring(1); // local form: 077… → 77…
    } else {
      national = s; // assume already national (7…)
    }

    if (!_national.hasMatch(national)) return null;
    return '+964$national';
  }

  static bool isValid(String input) => normalize(input) != null;
}

File: apps/admin/src/lib/iraqi-phone.ts
/**
 * Iraqi (+964) mobile phone handling — client-side mirror of
 * `services/api/src/common/phone.util.ts` (which is authoritative; this only
 * avoids an obviously-invalid round trip to the server before an OTP request).
 */
const E164_IRAQI_MOBILE = /^\+9647\d{9}$/;

export function normalizeIraqiPhone(input: string): string | null {
  if (typeof input !== "string") return null;

  let s = input.trim().replace(/[\s\-().]/g, "");
  if (s.startsWith("00")) s = "+" + s.slice(2);

  let national: string;
  if (s.startsWith("+964")) {
    national = s.slice(4);
  } else if (s.startsWith("964")) {
    national = s.slice(3);
  } else if (s.startsWith("0")) {
    national = s.slice(1);
  } else {
    national = s;
  }

  if (!/^7\d{9}$/.test(national)) return null;

  const e164 = "+964" + national;
  return E164_IRAQI_MOBILE.test(e164) ? e164 : null;
}
